node/src/components/small_network.rs

```diff
@@ -590,24 +647,29 @@ where
     ) -> Effects<Event<P>> {
         let _ = self.pending.remove(&peer_address);
 
+        let mut effects = Effects::new();
+
         if let Some(peer_id) = peer_id {
-            if let Some(err) = error {
+            if let Some(ref err) = error {
                 warn!(our_id=%self.our_id, %peer_id, %peer_address, %err, "outgoing connection failed");
             } else {
                 warn!(our_id=%self.our_id, %peer_id, %peer_address, "outgoing connection closed");
             }
-            return self.remove(effect_builder, &peer_id, false);
+            effects.extend(self.remove(effect_builder, &peer_id, false));
+        } else {
+            // If we are not calling remove, call the reconnection check explicitly.
+            effects.extend(self.reconnect_if_not_connected_to_any_known_addresses(effect_builder));
         }
 
         // If we don't have the node ID passed in here, it was never added as an
         // outgoing connection, hence no need to call `self.remove()`.
-        if let Some(err) = error {
+        if let Some(ref err) = error {
             warn!(our_id=%self.our_id, %peer_address, %err, "outgoing connection failed");
         } else {
             warn!(our_id=%self.our_id, %peer_address, "outgoing connection closed");
         }
 
-        Effects::new()
+        effects
```

Comment:
If you rearrange the code in this method and log at the beginning, you won't need to create the `mut effects` and just regurn from the `if`'s.

Reply:
Good catch! Although it would change the output order, this seems reasonable.

At the current point though I am loathe to make this kind of change for fear of accidentally introducing a regression. Can we maybe postpone this?

Reply:
Output order of what? Logs?